hw/ip/hmac/data/hmac.hjson

```diff
@@ -289,82 +383,86 @@
+        desc: '''Digest output.
 
+              If HMAC is disabled, the register shows result of SHA-2 256/384/512.
               Order of the digest is:
-              digest[255:0] = {DIGEST0, DIGEST1, DIGEST2, ... , DIGEST7};
+              digest[511:0] = {DIGEST0, DIGEST1, DIGEST2, ... , DIGEST15}. For SHA-2 256
+              DIGEST0, DIGEST2,....DIGEST14 are irrelevant/all-zero;
+              256-bit digest = {DIGEST1, DIGEST3, DIGEST5, DIGEST7, DIGEST9, DIGEST11, DIGEST13,
+              DIGEST15}. For SHA-2 384, DIGEST12-DIGEST15 are truncated and are all-zero.
 
               The digest gets cleared when `CFG.sha_en` transitions from 1 to 0.
               When `CFG.sha_en` is 0, these registers can be written by software.
               ''',
-        count: "NumWords",
+        count: "NumDigestWords",
         cname: "HMAC",
         swaccess: "rw",
         hwaccess: "hrw",
         hwext: "true",
         hwqe: "true",
         fields: [
-          { bits: "31:0", name: "digest", desc: "32-bit chunk of 256-bit Digest" }
+          { bits: "31:0", name: "digest", desc: "32-bit chunk of 512-bit digest" }
         ]
+        tags: [// Exclude digest from write-read checks because it does not write value
+               // to CSRs when digest_mode is SHA2_None.
+               "excl:CsrNonInitTests:CsrExclWriteCheck"]
+
       }
     }
-    { name: "MSG_LENGTH_LOWER",
-      desc: '''Received Message Length calculated by the HMAC in bits [31:0]
+    { multireg: {
+        name: "MSG_LENGTH",
+        desc: '''Received Message Length calculated by the HMAC in bits [127:0]
 
-            Message is byte granularity.
-            lower 3bits [2:0] are ignored.
+            Message is in byte granularity, thus lower 3bits [2:0] are ignored.
+            SHA-2 256 only assumes 64-bit length ({MSG_LENGTH2, MSG_LENGTH3}), so
+            {MSG_LENGTH0, MSG_LENGTH1} are irrelevant in this configuration.
 
             When `CFG.sha_en` is 0, this register can be written by software.
-            ''',
-      swaccess: "rw",
-      hwaccess: "hrw",
-      hwext: "true",
-      hwqe: "true",
-      fields: [
-        { bits: "31:0", name: "v", desc: "Message Length [31:0]" }
-      ]
-    }
-    { name: "MSG_LENGTH_UPPER",
-      desc: '''Received Message Length calculated by the HMAC in bits [63:32]
 
-            When `CFG.sha_en` is 0, this register can be written by software.
             ''',
-      swaccess: "rw",
-      hwaccess: "hrw",
-      hwext: "true",
-      hwqe: "true",
-      fields: [
-        { bits: "31:0", name: "v", desc: "Message Length [63:32]" }
-      ]
+        count: "4",
+        cname: "HMAC",
+        swaccess: "rw",
+        hwaccess: "hrw",
+        hwext: "true",
+        hwqe: "true",
+        fields: [
+          { bits: "31:0", name: "msg_length", desc: "32-bit of 128-bit message length" }
+        ]
+      }
     }
-    { skipto: "0x800" }
+    { skipto: "0x8000" }
```

Comment:
Do we need to leave so much space in the register map between CSRs and message FIFO? Wouldn't a skip of `0x1000` suffice (reserving 4 KiB for the CSRs)